core/src/main/scala/kafka/utils/CommandLineUtils.scala

```diff
@@ -36,16 +37,27 @@ object CommandLineUtils extends Logging {
     return commandOpts.args.length == 0 || commandOpts.options.has(commandOpts.helpOpt)
   }
 
+  def isPrintVersionNeeded(commandOpts: CommandDefaultOptions): Boolean = {
+    return commandOpts.options.has(commandOpts.versionOpt)
+  }
+
   /**
     * Check and print help message if there is no options or `--help` option
-    * from command line
+    * from command line, if `--version` is specified on the command line
+    * print version information and exit.
+    * NOTE: The function name is not strictly speaking correct anymore
+    * as it also checks whether the version needs to be printed, but
+    * refactoring this would have meant changing all command line tools
+    * and unnecessarily increased the blast radius of this change.
     *
     * @param commandOpts Acceptable options for a command
     * @param message     Message to display on successful check
     */
   def printHelpAndExitIfNeeded(commandOpts: CommandDefaultOptions, message: String) = {
     if (isPrintHelpNeeded(commandOpts))
       printUsageAndDie(commandOpts.parser, message)
+    if (isPrintVersionNeeded(commandOpts))
```

Comment:
Hmm.. I'm not sure this is called by all the commands go through `kafka-run-class.sh`. For example, `kafka.Kafka` does not seem to hit this path.

Reply:
I've been through the scripts in bin/ and looked at how many would be affected by this.  There are three classes that use CommandLineUtils but don't hit this method:  ReplicaVerificationTool, Kafka, StreamsResetter. 
Coincidentally for StreamsResetter this means that the "--help" option, while listed in the output still results in an error.

The issue with these classes is, I think, that they internally do not use an object that extends CommandDefaultOptions - hence ignore those options completely.

The rest of the tools either call this method or handle argument parsing in a totally different way.

Not sure what the correct way forward is, to be honest, command line parsing seems a little all over the place in general, so the only place where this is truly global is indeed in kafka-run-class.sh, but that doesn't feel right either, as it is a completely undocumented on the command line that way.

Reply:
If there only 3 exceptions, do you think it would be reasonable to implement `--version` separately for each of them? Perhaps in a follow-up we can consider how to refactor so that they all use `CommandDefaultOptions`. What do you think?

Reply:
Hi @hachikuji 
I am happy to implement those three separately for now.

In the interest of completion, handling those three would leave the following classes that do not support the version option (or help for that matter unless they implemented that separately):

- ConnectDistributed - Doesn't use CommandLine
- ConnectStandalone - Doesn't use CommandLine
- ProducerPerformance - Uses argparse4j instead
- VerifiableConsumer - Uses argparse4j instead
- VerifiableProducer - Uses argparse4j instead
- QuorumPeerMain - Zookeeper class
- ZooKeeperMain - Zookeeper class

I'm not sure why a small subset of these uses argparse4j, I briefly looked around but couldn't find a larger scheme to adopt this over the CommandLineUtils, do you know anything more about that?

Should we extend these classes as well to honor the version flag? Should be simple-ish, but still requires adding an option to all parsers, unless we create the equivalent of the CommandDefaultOptions for argparse4j classes. But I'd say this is better suited for the follow-up PR you mentioned to standardize command line parsing.

So, my proposal woud be:
add very simple implementations to ConnectDistributed and ConnectStandalone as well, leave the rest be for now and look at the larger parsing picture in a follow up PR.

Fair?

Reply:
Yeah, I'm not sure why they use argparse4j. Maybe just to make our lives harder. Your suggestion sounds fine to me. Would you mind filing JIRAs for the remaining work?

Reply:
Happy to!
I'll kick off a thread on the mailing list first though, to ask around what people think we should do, personally, I'd like us to use just one command line parser, don't really care which one, but two is one too many for my liking.
But maybe there were reasons for having two.

Reply:
I suspect there are no good reasons, but starting a discussion is a good idea. There is an ancient KIP which was attempting to consolidate these tools: https://cwiki.apache.org/confluence/display/KAFKA/KIP-14+-+Tools+Standardization. If you have any interest, it would really be helpful for someone to pick that up again. I think one of the reasons for the naming inconsistencies has been the fact that we are using different tools.

Reply:
Having looked some more at the two Connect classes, there is a small catch: everything we'd need to add the --version parameter to those classes is in the core Kafka project which Connect currently does not declare as a dependency. It seems a bit excessive to add that just for this purpose. 
So I'll reword my suggestion to ignore these two for now and come up with an overall action plan first, before making any larger changes.